File: framework/ui-elements/src/displays/components/BaseNumber.js
import AbstractBaseNumber from "../../../gen/displays/components/AbstractBaseNumber";
import {normalizeHighlight, withHighlight, withHighlightBackground} from "./Highlight";

export default class BaseNumber extends AbstractBaseNumber {

	constructor(props) {
		super(props);
		this.state = {
			error: null,
			value : this.props.value,
		    decimals: this.props.decimals != null ? this.props.decimals : 0,
			expanded : this.props.expanded,
			style: this.props.style,
			prefix : this.props.prefix,
			suffix : this.props.suffix,
			...this.state,
			highlighted: normalizeHighlight(this.props.highlighted),
		}
	};

	refreshHighlight = (highlighted) => {
		this.setState({ highlighted: normalizeHighlight(highlighted, this.props.highlighted) });
	};

	style() {
		return super.style();
	};

	highlightStyle() {
		return withHighlight({}, this.state.highlighted);
	};

	highlightBackgroundStyle() {
		return withHighlightBackground({}, this.state.highlighted);
	};

	refresh = (value) => {
	};

    refreshError = (value) => {
		this.setState({ error: value });
    };

	refreshExpanded = (value) => {
		this.setState({expanded : value});
	};

	refreshDecimals = (value) => {
		this.setState({decimals : value});
	};

	refreshPrefix = (value) => {
		this.setState({prefix : value});
	};

	refreshSuffix = (value) => {
		this.setState({suffix : value});
	};

	refreshStyle = (value) => {
		this.setState({style : value});
	};
}

File: framework/ui/src/io/intino/alexandria/ui/displays/components/BaseNumber.java
package io.intino.alexandria.ui.displays.components;

import io.intino.alexandria.core.Box;
import io.intino.alexandria.schemas.Highlight;
import io.intino.alexandria.ui.displays.notifiers.BaseNumberNotifier;

public class BaseNumber<DN extends BaseNumberNotifier, B extends Box> extends AbstractBaseNumber<DN, B> {
    private boolean expanded;

    public BaseNumber(B box) {
        super(box);
    }

    public enum Style { Number, Currency, Bytes, Percentage, Exponential }

    public BaseNumber<DN, B> expanded(boolean value) {
        this.expanded = value;
        notifier.refreshExpanded(value);
        return this;
    }

    public BaseNumber<DN, B> countDecimals(int count) {
        notifier.refreshDecimals(count);
        return this;
    }

    public boolean expanded() {
        return expanded;
    }

    public BaseNumber<DN, B> prefix(String prefix) {
        notifier.refreshSuffix(prefix);
        return this;
    }

    public BaseNumber<DN, B> suffix(String suffix) {
        notifier.refreshSuffix(suffix);
        return this;
    }

    public BaseNumber<DN, B> error(String error) {
        notifier.refreshError(error);
        return this;
    }

    public BaseNumber<DN, B> highlight(String textColor, String backgroundColor) {
        notifier.refreshHighlight(new Highlight().textColor(textColor).backgroundColor(backgroundColor));
        return this;
    }

    public BaseNumber<DN, B> highlight(String color) {
        return highlight(null, color);
    }

    public BaseNumber<DN, B> clearHighlight() {
        return highlight(null, null);
    }

    public BaseNumber<DN, B> style(Style style) {
        notifier.refreshStyle(style.name());
        return this;
    }
}
